qa/workunits/rbd/krbd_exclusive_option.sh

```diff
@@ -84,79 +109,107 @@ sudo rbd unmap $DEV
 dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
 assert_locked $DEV
-OTHER_DEV=$(sudo rbd map -o noshare,exclusive $IMAGE_NAME)
+dd if=/dev/urandom of=$OTHER_DEV bs=4k count=10 oflag=direct
 assert_locked $OTHER_DEV
 sudo rbd unmap $DEV
 sudo rbd unmap $OTHER_DEV
 assert_unlocked
 
-DEV=$(sudo rbd map -o exclusive $IMAGE_NAME)
+DEV=$(sudo rbd map $IMAGE_NAME)
 assert_locked $DEV
-expect_false sudo rbd map -o noshare,exclusive $IMAGE_NAME
+OTHER_DEV=$(sudo rbd map -o noshare,exclusive $IMAGE_NAME)
+assert_locked $OTHER_DEV
+dd if=$DEV of=/dev/null bs=4k count=10 iflag=direct
+expect_false dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
+assert_locked $OTHER_DEV
+sudo rbd unmap $OTHER_DEV
+assert_unlocked
+dd if=$DEV of=/dev/null bs=4k count=10 iflag=direct
+assert_unlocked
+dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
+assert_locked $DEV
+sudo rbd unmap $DEV
+assert_unlocked
+
+DEV=$(sudo rbd map -o lock_on_read $IMAGE_NAME)
+assert_locked $DEV
+OTHER_DEV=$(sudo rbd map -o noshare,exclusive $IMAGE_NAME)
+assert_locked $OTHER_DEV
+expect_false dd if=$DEV of=/dev/null bs=4k count=10 iflag=direct
+expect_false dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
+assert_locked $OTHER_DEV
+sudo rbd unmap $OTHER_DEV
+assert_unlocked
+dd if=$DEV of=/dev/null bs=4k count=10 iflag=direct
+assert_locked $DEV
+dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
 assert_locked $DEV
 sudo rbd unmap $DEV
 assert_unlocked
 
 DEV=$(sudo rbd map -o exclusive $IMAGE_NAME)
 assert_locked $DEV
-OTHER_DEV=$(sudo rbd map -o noshare $IMAGE_NAME)
-dd if=/dev/urandom of=$OTHER_DEV bs=4k count=10 oflag=direct &
-PID=$!
-sleep 20
+expect_false sudo rbd map -o noshare $IMAGE_NAME
 assert_locked $DEV
-[[ "$(ps -o stat= $PID)" =~ ^D ]]
 sudo rbd unmap $DEV
-wait $PID
-assert_locked $OTHER_DEV
-sudo rbd unmap $OTHER_DEV
 assert_unlocked
 
 DEV=$(sudo rbd map -o exclusive $IMAGE_NAME)
 assert_locked $DEV
-OTHER_DEV=$(sudo rbd map -o noshare,lock_timeout=60 $IMAGE_NAME)
-dd if=/dev/urandom of=$OTHER_DEV bs=4k count=10 oflag=direct &
-PID=$!
-sleep 20
+expect_false sudo rbd map -o noshare,exclusive $IMAGE_NAME
 assert_locked $DEV
-[[ "$(ps -o stat= $PID)" =~ ^D ]]
-expect_false wait $PID
+sudo rbd unmap $DEV
+assert_unlocked
+
+DEV=$(sudo rbd map $IMAGE_NAME)
 assert_locked $DEV
+rbd resize --size 1G $IMAGE_NAME
+assert_unlocked
 sudo rbd unmap $DEV
-sudo rbd unmap $OTHER_DEV
 assert_unlocked
 
 DEV=$(sudo rbd map -o exclusive $IMAGE_NAME)
 assert_locked $DEV
-sudo rbd map -o noshare,lock_on_read $IMAGE_NAME &
-SUDO_PID=$!
-sleep 20
+expect_false rbd resize --size 2G $IMAGE_NAME
 assert_locked $DEV
-PID="$(ps -o pid= --ppid $SUDO_PID)"
-[[ "$(ps -o stat= $PID)" =~ ^D ]]
 sudo rbd unmap $DEV
-wait $SUDO_PID
+assert_unlocked
+
+DEV=$(sudo rbd map $IMAGE_NAME)
+assert_locked $DEV
+dd if=/dev/urandom of=$DEV bs=4k count=10 oflag=direct
+{ sleep 10; blacklist_add $DEV; } &
+PID=$!
+expect_false dd if=/dev/urandom of=$DEV bs=4k count=200000 oflag=direct
```

Comment:
This seems potentially racy. You would need to ensure that the OSD has the updated OSD map w/ the blacklist entry for this to actually fail.

Reply:
While it is racy in theory, this test case is doing 200000 I/Os at QD=1 with messenger failure injection enabled.  Uninterrupted it runs for at least a few minutes, so there should be more than enough time for the blacklist entry to propagate.

A CentOS 7 run turned up a udev-related race in a different test case, I'm pushing a fix.